src/linker/Linker/MessageContainer.cs

```diff
@@ -44,14 +46,37 @@ namespace Mono.Linker
 		/// <param name="subcategory">Optionally, further categorize this error</param>
 		/// <param name="origin">Filename, line, and column where the error was found</param>
 		/// <returns>New MessageContainer of 'Error' category</returns>
-		public static MessageContainer CreateErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null)
+		internal static MessageContainer CreateErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null)
 		{
 			if (!(code >= 1000 && code <= 2000))
 				throw new ArgumentException ($"The provided code '{code}' does not fall into the error category, which is in the range of 1000 to 2000 (inclusive).");
 
 			return new MessageContainer (MessageCategory.Error, text, code, subcategory, origin);
 		}
 
+		/// <summary>
+		/// Create a custom error message.
+		/// </summary>
+		/// <param name="text">Humanly readable message describing the error</param>
+		/// <param name="code">A custom error ID. This code should be greater than or equal to 6001
+		/// to avoid any collisions with existing and future linker errors</param>
+		/// <param name="subcategory">Optionally, further categorize this error</param>
+		/// <param name="origin">Filename or member where the error is coming from</param>
+		/// <returns>Custom MessageContainer of 'Error' category</returns>
+		public static MessageContainer CreateCustomErrorMessage (string text, int code, string subcategory = MessageSubCategory.None, MessageOrigin? origin = null)
+		{
+			var callingAssembly = Assembly.GetCallingAssembly ();
+#if DEBUG
+			Debug.Assert (!string.Equals (callingAssembly.GetName ().Name, Constants.ILLink, StringComparison.OrdinalIgnoreCase),
+				"Internal error: 'CreateCustomErrorMessage' is intended to be used by external assemblies only. Use 'CreateErrorMessage' instead.");
+#endif
+			if ((typeof(MessageContainer).Assembly != callingAssembly) && code <= 600)
```

Comment:
Why do we need the calling assembly check here. The assert above should make sure that linker itself never calls this, so we can get here only from custom steps. So simply check the code is above 6000 and that should do it.